### PR TITLE
fix(daemon): graceful SIGTERM shutdown

### PR DESCRIPTION
**Pull request by aaf2tbz**
## Summary

Fixes five interacting bugs in daemon shutdown that caused stale connector state, phantom session claims, and post-teardown crashes when the daemon received SIGTERM/SIGINT.

### What changed

**3 files, 104 additions, 16 deletions:**

- **`daemon.ts`** — Rewrote `cleanup()` with phased shutdown:
  - **Phase 0**: Capture `serve()` return value → close HTTP server with 15s drain timeout + `closeAllConnections` fallback, so in-flight `session-end` hooks complete before state teardown
  - **Phase 1**: Clear `commitTimer` (5s debounce) and `syncTimer` (2s debounce) immediately, reset `commitPending`/`syncPending` flags — prevents callbacks from firing after `closeDbAccessor()`
  - **Phase 2**: Bulk `releaseAllSessions()` + `clearAllPresence()` before DB close — connectors see clean state immediately instead of waiting 4h stale expiry
  - Made `stopGitSyncTimer()` async — tracks `gitSyncPromise` and awaits in-flight git push/pull before DB teardown

- **`session-tracker.ts`** — Added `releaseAllSessions()` for bulk session cleanup on shutdown

- **`cross-agent.ts`** — Added `clearAllPresence()` for bulk presence cleanup on shutdown

### Why

| Bug | Symptom | Root cause |
|-----|---------|------------|
| 1 | Crash on closed DB after SIGTERM | `commitTimer`/`syncTimer` callbacks fire after `closeDbAccessor()` |
| 2 | Corrupted git state | In-flight `gitSync()` races against DB teardown |
| 3 | Stale session claims for 4 hours | HTTP server not drained — `session-end` hooks abort, `releaseSession`/`removeAgentPresence` never fire |
| 4 | Phantom peers in cross-agent view | Sessions/presence not bulk-released on shutdown |
| 5 | Shutdown hangs on keep-alive | HTTP server drain had no timeout |

### Edge cases considered

- `closeAllConnections` is Node 18.2+ only — guarded with `"closeAllConnections" in srv` runtime check
- `gitSyncPromise` race: uses identity comparison (`gitSyncPromise === work`) to avoid clearing a newer sync's promise
- Drain timeout (15s) balances letting in-flight requests finish vs. not blocking shutdown indefinitely
- `commitPending`/`syncPending` reset to `false` so no stale guard flags survive into the next debounce cycle edge case

## Test plan

- [x] `bun run build` — passes
- [x] `bun run typecheck` — no new errors (pre-existing only)
- [x] `bun test` — 783 tests pass (1805 expects)
- [ ] Manual: start daemon, trigger file watcher commit debounce, send SIGTERM during 15s window — verify no DB crash
- [ ] Manual: start daemon with active connector session, send SIGTERM — verify session claims clear immediately (not 4h later)
- [ ] Manual: send SIGTERM during active git sync — verify sync completes before process exits

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by aaf2tbz** [comment]
Hi @aaf2tbz - I'm an automated code reviewer powered by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer). I'm taking a look at the fixes in `fix(daemon): graceful SIGTERM shutdown` (commit `ef2cdb10`) now and I'll follow up shortly with feedback.

File: packages/daemon/src/daemon.ts [hunk 1]
```diff
@@ -10345,8 +10376,51 @@ async function importExistingMemoryFiles(): Promise<number> {
 // ============================================================================
```
Comment (aaf2tbz) [inline packages/daemon/src/daemon.ts:10378]: **Rust parity rule violation.** AGENTS.md states: 'Any behavioral change made to @signet/daemon must also be reflected in packages/daemon-rs/.' This PR rewrites the shutdown sequence — phased cleanup(), async stopGitSyncTimer(), gitSyncPromise tracking, and bulk session/presence release — but the diff includes zero changes to packages/daemon-rs/. The shadow daemon will diverge and continue exhibiting the pre-fix crash-on-SIGTERM behavior. Per the parity rule this must be reflected before merge.

File: packages/daemon/src/daemon.ts [hunk 2]
```diff
@@ -10345,8 +10376,51 @@ async function importExistingMemoryFiles(): Promise<number> {
...
+	// Phase 0: Stop accepting new requests — close HTTP server first so
+	// in-flight session-end hooks can still complete before we tear down
+	// state, but no new requests arrive during shutdown. The 15s timeout
+	// exceeds OpenClaw's WRITE_TIMEOUT (10s) so session-end hooks can
+	// finish and respond before we force-close.
+	// ------------------------------------------------------------------
+	if (httpServer) {
+		const srv = httpServer;
+		await new Promise<void>((resolve) => {
+			const timeout = setTimeout(() => {
```
Comment (aaf2tbz) [inline packages/daemon/src/daemon.ts:10394]: PR description says 'Drain timeout (5s)' in both the summary table and the edge-cases section, but the code uses `15_000` ms (15 seconds). The inline comment correctly justifies 15s (OpenClaw WRITE_TIMEOUT is 10s), so the code is likely correct, but the description is wrong. Reviewers approving based on the stated 5s will have an incorrect mental model of the shutdown delay budget. Update the PR description to match.

File: packages/daemon/src/cross-agent.ts [hunk 3]
```diff
@@ -777,6 +777,13 @@ export async function relayMessageViaAcp(request: AcpRelayRequest): Promise<AcpR
 	}
 }
 
+/** Remove all presence entries (for graceful shutdown). Returns count cleared. */
+export function clearAllPresence(): number {
```
Comment (aaf2tbz) [inline packages/daemon/src/cross-agent.ts:781]: `clearAllPresence()` calls `presenceByKey.clear()` without emitting a `presence:remove` event for each cleared entry, unlike `removeAgentPresence()` which calls `emit({type: 'presence', action: 'remove', ...})`. During shutdown this is safe in practice — SSE connections are already closed after Phase 0, and no new subscribers should be active. However, it silently breaks the emit contract for any in-process subscriber tracking presence counts. At minimum add a comment explaining why emission is intentionally skipped (e.g. 'server already closed, no live subscribers'). Compare to `resetCrossAgentStateForTest()` which also skips emit — that precedent exists but is test-only.

**Review by aaf2tbz — COMMENTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: claude-sonnet-4-6 | commit: `ef2cdb10`

[Automated PR Review — pr-reviewer bot] Found one blocking issue (Rust parity rule violation) and two warnings (description/code timeout mismatch, missing subscriber events in bulk presence clear). The phased shutdown logic is sound and addresses the stated bugs correctly.

**Confidence:** High - The diff is well-scoped to three files with clear, linear control flow. The Rust parity violation is unambiguous given AGENTS.md requirements. The 5s vs 15s timeout discrepancy is directly readable in the diff. The missing emit() in clearAllPresence() is verifiable by comparing it to removeAgentPresence() in the same file.

**Comment by PR-Reviewer-Ant** [comment]
Hi @aaf2tbz - I'm an automated code reviewer powered by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer). I'm taking a look at the fixes in `fix(daemon): graceful SIGTERM shutdown` (commit `ef2cdb10`) now and I'll follow up shortly with feedback.

File: packages/daemon/src/cross-agent.ts [hunk 4]
```diff
@@ -777,6 +777,13 @@ export async function relayMessageViaAcp(request: AcpRelayRequest): Promise<AcpR
...
+	const count = presenceByKey.size;
+	presenceByKey.clear();
+	return count;
```
Comment (PR-Reviewer-Ant) [inline packages/daemon/src/cross-agent.ts:795]: `clearAllPresence()` calls `presenceByKey.clear()` directly without emitting `{type: "presence", action: "remove"}` events to subscribers. Compare `removeAgentPresence()`, which emits per-entry removal events. Any SSE/streaming clients subscribed via `subscribeCrossAgentEvents()` will observe a silent Map clear rather than proper removal notifications. During shutdown, connections are closing anyway (Phase 0 drains the HTTP server), so the practical impact is low — but clients that handle `remove` events gracefully (e.g. to update UI state or trigger reconnect logic) will instead see a connection reset. If this is intentional for shutdown performance, add a comment noting the deliberate event-skip.

File: packages/daemon/src/daemon.ts [hunk 5]
```diff
@@ -9469,27 +9485,42 @@ function startGitSyncTimer() {
...
 
-		logger.debug("git", "Running periodic sync...");
-		const result = await gitSync();
-		if (!result.success) {
-			logger.warn("git", `Periodic sync failed: ${result.message}`);
-		}
+			logger.debug("git", "Running periodic sync...");
+			const result = await gitSync();
+			if (!result.success) {
+				logger.warn("git", `Periodic sync failed: ${result.message}`);
```
Comment (PR-Reviewer-Ant) [inline packages/daemon/src/daemon.ts:9500]: `work.finally(cb)` produces a derived promise that is never awaited or caught. If `work` rejects during a normal interval tick (not shutdown), both `work` itself and the `.finally()` chain promise are unhandled. `stopGitSyncTimer()` catches `gitSyncPromise` (which is `work`) only at shutdown — but `.finally()` creates a second, independent rejected promise that is never handled. Node.js/Bun will emit `unhandledRejection` for it. The original `setInterval(async () => {...})` pattern also had one unhandled rejection; this introduces a second one via the `.finally()` chain. Fix: attach `.catch(() => {})` to `work` or to the `.finally()` result, since both paths are intentionally best-effort.

File: packages/daemon/src/daemon.ts [hunk 6]
```diff
@@ -10345,8 +10376,51 @@ async function importExistingMemoryFiles(): Promise<number> {
...
+	// state, but no new requests arrive during shutdown. The 15s timeout
+	// exceeds OpenClaw's WRITE_TIMEOUT (10s) so session-end hooks can
+	// finish and respond before we force-close.
+	// ------------------------------------------------------------------
+	if (httpServer) {
+		const srv = httpServer;
+		await new Promise<void>((resolve) => {
+			const timeout = setTimeout(() => {
+				logger.warn("daemon", "HTTP server drain timed out, forcing close");
+				if ("closeAllConnections" in srv && typeof srv.closeAllConnections === "function") {
```
Comment (PR-Reviewer-Ant) [inline packages/daemon/src/daemon.ts:10396]: `srv.close(callback)` is invoked without an existence/type guard. In the same block, `closeAllConnections` is correctly guarded: `"closeAllConnections" in srv && typeof srv.closeAllConnections === "function"`. There is no analogous guard for `.close()`. If at runtime the server object does not expose `.close()` (e.g. a Bun-native server that uses `.stop()` instead), this throws synchronously inside the `new Promise<void>()` constructor, rejecting the promise. `cleanup()` does not wrap Phase 0 in a try/catch, so this rejection would propagate up and potentially abort the rest of shutdown — DB close, timer cleanup, session release — leaving state inconsistent. The asymmetry with the `closeAllConnections` guard pattern is the primary concern. Mitigation: add `"close" in srv && typeof srv.close === "function"` guard, or wrap Phase 0 in try/catch with a fallback.

File: packages/daemon/src/daemon.ts [hunk 7]
```diff
@@ -10345,8 +10376,51 @@ async function importExistingMemoryFiles(): Promise<number> {
...
+					srv.closeAllConnections();
```
Comment (PR-Reviewer-Ant) [inline packages/daemon/src/daemon.ts:10397]: PR description says "5s drain timeout" but the implementation uses `15_000` (15 seconds). The inline code comment correctly justifies 15s (must exceed OpenClaw's 10s `WRITE_TIMEOUT`), but the PR description is stale and contradicts the implementation. Low impact, but can mislead reviewers or future bisects.

**Review by PR-Reviewer-Ant — COMMENTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: claude-sonnet-4-6 | commit: `ef2cdb10`

[Automated Bot Review — pr-reviewer] Reviewed fix(daemon): graceful SIGTERM shutdown (#307). The phased shutdown design is sound and addresses the stated bugs. Found three warnings: bulk presence clear bypasses event emission, a new unhandled-rejection surface in the git sync IIFE pattern, and an unguarded `srv.close()` call (unlike the guarded `closeAllConnections` check in the same block). One nitpick: PR description claims 5s drain timeout but code uses 15s.

**Confidence:** Medium - Shutdown and concurrent async paths cannot be fully verified statically. The logic reads correctly but runtime behavior (unhandled rejections, server drain semantics) requires integration testing to confirm. The typecheck passing gives confidence on the srv.close() concern.

**Review by PR-Reviewer-Ant — COMMENTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: claude-sonnet-4-6 | commit: `2d759f53`

[Automated Bot Review — pr-reviewer] Incremental review of 2d759f53 over prior-reviewed ef2cdb10. Both substantive prior findings are resolved: event emission is now present in `clearAllPresence()` and the git-sync IIFE unhandled-rejection surface is closed with a `.catch()` handler. No new issues found.

**Confidence:** High - The incremental diff is small (two targeted fixes) and directly addresses the two actionable prior findings. The `activeCount` arithmetic in the new loop is correct (presenceByKey.size is constant during the loop, so `size - count - 1` yields the post-removal count for each step, matching the convention in removeAgentPresence). The srv.close() concern from the prior review was a false positive — server.close() is a baseline Node/Bun API and does not need a version guard; only closeAllConnections required one.

**Review by PR-Reviewer-Ant — COMMENTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: claude-sonnet-4-6 | commit: `1f674844`

[Automated Bot Review — pr-reviewer] Incremental review of 1f674844 over prior-reviewed 2d759f53. All prior findings are resolved. This increment is documentation-only (workspace path canonicalization across docs). Two minor doc quality issues found.

**Confidence:** High - The incremental diff touches only documentation files. No code logic, types, or runtime behavior changed. The two findings are documentation clarity issues, not correctness bugs.


Unmapped findings (not on changed lines):
- docs/BENCHMARKS.md:44 - Unquoted `$SIGNET_WORKSPACE` in shell commands — if the workspace path contains spaces or special characters these `cp` and `ln -sf` invocations will break silently. Should be quoted: `"$SIGNET_WORKSPACE"`.
- docs/CLI.md:781 - The `SIGNET_PATH` default column now reads `$SIGNET_WORKSPACE`, which is tautological — `$SIGNET_WORKSPACE` is the resolved value *of* `SIGNET_PATH`, not a separate default. The original `~/.agents` was clearer. Users reading this table in isolation will not understand what the default is. Consider `~/.agents` or a literal like `(active workspace, default ~/.agents)`.

**Review by PR-Reviewer-Ant — COMMENTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: claude-sonnet-4-6 | commit: `adcd4c11`

Harness output could not be parsed as structured JSON.

```pr-review-json
{
  "summary": "[Automated Bot Review — pr-reviewer] Incremental review of adcd4c11 over prior-reviewed 2d759f53. This diff is entirely documentation and changelog — no code changes. The two prior substantive findings were already resolved. Two self-referential documentation errors introduced in docs/CLI.md are worth flagging.",
  "verdict": "comment",
  "confidence": {
    "level": "high",
    "justification": "The incremental diff touches only markdown documentation files. No runtime code changed. The issues found are clear documentation errors (circular variable references) that are unambiguous."
  },
  "ui_screenshot_needed": false,
  "comments": [
    {
      "file": "docs/CLI.md",
      "line": 783,
      "body": "Documentation regression: the `SIGNET_PATH` env-var row now shows its default as `$SIGNET_WORKSPACE`, which is self-referential — `SIGNET_PATH` is the env var that *defines* the workspace path. The prior text `~/.agents` was correct. Readers of this table cannot infer the actual default without already knowing it. README.md adds a clarifying note ("Default is `~/.agents`, configurable via `signet workspace set <path>`"); the same note (or the concrete default) should appear here.",
      "severity": "warning"
    },
    {
      "file": "docs/CLI.md",
      "line": 118,
      "body": "`-p, --path <path>` option description now reads `(default: $SIGNET_WORKSPACE)`, which has the same circularity: the flag sets the workspace path, so documenting its default as the workspace path is uninformative. Was `~/.agents`, which was concrete and correct.",
      "severity": "warning"
    },
    {
      "file": "CHANGELOG.md",
      "line": 4,
      "body": "CHANGELOG entries for PRs #302 and #305 are added here, but the fix introduced by this PR (#307 — graceful SIGTERM shutdown) has no entry. If changelog maintenance is automated or deferred, this is expected; otherwise it looks like an omission.",
      "severity": "nitpick"
    },
    {
      "file": "AGENTS.md",
      "line": 1,
      "body": "These are AGENTS.md / doc-only changes bundled into a daemon shutdown fix PR. The changes are correct in isolation (replacing hardcoded `~/.agents` with `$SIGNET_WORKSPACE`), but they belong to the workspace-path feature (PR #302), not to this fix. Not a correctness issue, but worth noting for traceability.",
      "severity": "nitpick"
    }
  ]
}
```

**Review by PR-Reviewer-Ant — COMMENTED** [review]
> Automated review by [pr-reviewer](https://github.com/NicholaiVogel/pr-reviewer) | model: claude-sonnet-4-6 | commit: `aef530f0`

[Automated Bot Review — pr-reviewer] Incremental review of aef530f over prior-reviewed adcd4c11. This increment adds a single 1,057-line competitive-systems research document. All prior code findings remain resolved. One scope concern flagged.

**Confidence:** High - The incremental diff touches only one new documentation file. No runtime code changed. The finding is unambiguous — the research document is structurally valid per the specs pipeline rules but is unrelated to the PR's stated purpose.


Unmapped findings (not on changed lines):
- docs/research/technical/RESEARCH-COMPETITIVE-SYSTEMS.md:1 - This research document is unrelated to the PR's stated purpose ('fix(daemon): graceful SIGTERM shutdown'). The PR description explicitly enumerates '3 files, 104 additions, 16 deletions' covering daemon.ts, session-tracker.ts, and cross-agent.ts — it makes no mention of a competitive-systems research artifact. This looks like an accidental inclusion from a diverged branch, or work that belongs in a separate PR. Worth confirming the commit is intentional here before merging.
